### PR TITLE
Changes pertaining to supporting 0.7

### PR DESCRIPTION
**Pull request by sambitdash**

### CODE REVIEW

File: src/slice.jl [hunk 1]
```diff
@@ -71,14 +72,8 @@ end
...
-        # TODO: This should work but won't. Fix this in Julia.
-        #copy!(s.data, i+1, s.data, i, s.n - i + 1)
-        if isbits(T)
-            unsafe_copy!(pointer(s.data, i+1), pointer(s.data, i), s.n - i + 1)
-        else
-            for k in 0:(s.n - i)
-                @inbounds s.data[s.n+1-k] = s.data[s.n-k]
-            end
+        for k in 0:(s.n - i)
+            @inbounds s.data[s.n+1-k] = s.data[s.n-k]
```
Comment (ScottPJones) [inline src/slice.jl]: Why did you remove the fast path (using unsafe_copy! for isbits types)?

File: src/slice.jl [hunk 2]
```diff
@@ -121,13 +116,8 @@ end
...
     if count < length(xs) && 1 <= i <= count + 1
-        if isbits(T)
-            unsafe_copy!(pointer(xs, i+1),
-                         pointer(xs, i), count - i + 1)
-        else
-            for k in 0:(count - i)
-                @inbounds xs[count+1-k] = xs[count-k]
-            end
+        for k in 0:(count - i)
+            @inbounds xs[count+1-k] = xs[count-k]
```
Comment (ScottPJones) [inline src/slice.jl]: Same here, did the fast path not work in v0.7?  I think it really just needed a name change (which Compat would make work still in v0.6.2), from `unsafe_copy!` to `unsafe_copyto!`

File: test/runtests.jl [hunk 3]
```diff
@@ -99,28 +105,25 @@ end
...
-function validparents(node::IntervalTrees.LeafNode)
-    return true
-end
-
+validparents(node::IntervalTrees.LeafNode) = true
 
 # Verify that sibling/cousin pointers are correct
 function validsiblings(t::IntervalTrees.IntervalBTree)
-    if !isnull(t.root.left) || !isnull(t.root.right)
+    if !(t.root.left === nothing) || !(t.root.right === nothing)
```
Comment (ScottPJones) [inline test/runtests.jl]: Could be a lot simpler as:
`t.root.left === t.root.right === nothing || return false`

File: src/slice.jl (hunk 1 above)
Comment (sambitdash) [reply src/slice.jl]: I found it rather slowing down the test runs and allocating 1K bytes per each allocation and 10+ms for every execution. The original test case pass time was coming down 50 times slower than 0.6. Hence, had to eliminate additional allocations and calls that were contributing to that.

File: src/slice.jl (hunk 2 above)
Comment (sambitdash) [reply src/slice.jl]: As per the previous comment.

File: test/runtests.jl (hunk 3 above)
Comment (sambitdash) [reply test/runtests.jl]: agreed will do the change. One of the reasons I did not change the code significantly was because I felt the person who wrote it may not be able to review it and others may object to significant logical changes.

File: src/slice.jl (hunk 1 above)
Comment (ScottPJones) [reply src/slice.jl]: Had you tried changing the name to `unsafe_copyto!`?
The slowdown may just have been because of the deprecation of `unsafe_copy!`
(deprecated code, even though they print out the deprecation warning only once, slow down the code substantially for all calls)

Comment (sambitdash) [reply src/slice.jl]: Will try that. My normal interpretation of deprecation has been it's will not be maintained so don't use but if used it would do no harm. But I see in Julia many places deprecated code gets so much performance hit that it gets to the extent of being unusable.

File: src/IntervalTrees.jl [hunk 4]
```diff
@@ -602,7 +598,7 @@ function _push!(t::IntervalBTree{K, V, B},
                 node::InternalNode{K, V, B},
                 entry::V, unique_key::Bool, update::Bool) where {K, V, B}
     i = findidx(node, entry) # key index
-    j = i <= length(node) - 1 && entry >= node.keys[i] ? i + 1 : i # child index
+    j = i <= length(node) - 1 && entry < node.keys[i] ? i : i + 1 # child index
```
Comment (sambitdash) [inline src/IntervalTrees.jl]: Please review this code in light of #36. While this is just a workaround a complete elaborate fix will needed as described in #36.

File: src/IntervalTrees.jl [hunk 5]
```diff
@@ -641,11 +637,11 @@ function _push!(t::IntervalBTree{K, V, B},
...
         rightnode = rightleaf
         hassplit = true
-        while !isnull(parent)
-            p = get(parent)
+        while parent !== nothing
+            p = parent
             if hassplit
                 i = findidx(p, entry) # key index
-                j = i <= length(p) - 1 && entry >= p.keys[i] ? i + 1 : i # child index
+                j = i <= length(p) - 1 && entry < p.keys[i] ? i : i + 1 # child index
```
Comment (sambitdash) [inline src/IntervalTrees.jl]: Please review this code in light of #36. While this is just a workaround a complete elaborate fix will needed as described in #36.

File: test/runtests.jl [hunk 6]
```diff
@@ -79,7 +83,9 @@ function validkeys(node::IntervalTrees.LeafNode, minint, maxint)
             return false
         end
 
-        if !(minint <= entry <= maxint) || last(entry) > node.maxend
+        interval = IntervalTrees.Interval(first(entry), last(entry))
+
+        if !(minint <= interval <= maxint) || last(interval) > node.maxend
```
Comment (sambitdash) [inline test/runtests.jl]: Please review this code in light of #36. While this is just a workaround a complete elaborate fix will needed as described in #36.

File: src/IntervalTrees.jl [hunk 7]
```diff
@@ -1018,7 +1022,7 @@ end
 
 function Base.haskey(t::InternalNode{K, V, B}, key::AbstractInterval{K}) where {K, V, B}
     i = findidx(t, key)
-    if i <= length(t) - 1 && key >= t.keys[i]
+    if i <= length(t) - 1 && !(key < t.keys[i])
```
Comment (sambitdash) [inline src/IntervalTrees.jl]: Please review this code in light of #36. While this is just a workaround a complete elaborate fix will needed as described in #36.

File: src/IntervalTrees.jl [hunk 8]
```diff
@@ -1038,22 +1042,22 @@ function Base.findfirst(t::LeafNode{K, V, B}, key::AbstractInterval{K}, f) where
...
 
-    return Nullable{V}()
+    return nothing
 end
 
 
 function Base.findfirst(t::InternalNode{K, V, B}, key::AbstractInterval{K}, f) where {K, V, B}
     i = findidx(t, key)
-    if i <= length(t) - 1 && key >= t.keys[i]
+    if i <= length(t) - 1 && !(key < t.keys[i])
```
Comment (sambitdash) [inline src/IntervalTrees.jl]: Please review this code in light of #36. While this is just a workaround a complete elaborate fix will needed as described in #36.

File: src/slice.jl (hunk 1 above)
Comment (sambitdash) [reply src/slice.jl]: https://github.com/BioJulia/IntervalTrees.jl/pull/37/commits/bfe59b002e5f687d99faf5a65c1435efface8379 has the changes as suggested.

File: src/slice.jl (hunk 2 above)
Comment (sambitdash) [reply src/slice.jl]: https://github.com/BioJulia/IntervalTrees.jl/pull/37/commits/bfe59b002e5f687d99faf5a65c1435efface8379 has the changes as suggested.

File: test/runtests.jl (hunk 3 above)
Comment (sambitdash) [reply test/runtests.jl]: https://github.com/BioJulia/IntervalTrees.jl/pull/37/commits/bfe59b002e5f687d99faf5a65c1435efface8379 has the changes as suggested.

**Comment by ScottPJones** [comment]
The changes for dealing with the `isequal`/`==`/`>=` etc. changes seem fine to me.
I sympathize, I myself am having to deal with the changes to `===` for strings.

**Comment by sambitdash** [comment]
@ScottPJones I have no concerns as I know the current workarounds will work for now. But architectural decisions have to be taken for #36 by the maintainers on whatever may be best for the 3 types:

`AbstractInterval{K}, Interval{K}, IntervalValue{K, V}` in terms of the `==` and `hash` operator if `Dict` datastructure needs to be used.

**Comment by sambitdash** [comment]
The MacOS X build has failed due to a networking error not a real failure due to code changes.

**Comment by ScottPJones** [comment]
Looks good to me now - it should probably have the commits squashed (although I believe they can be squashed during the merge process)

**Comment by sambitdash** [comment]
It's just a button click while merging :-).

**Comment by bicycle1885** [comment]
The overall changes look reasonable to me. However, in my quick benchmarks, I found significant performance degradation (~4-5x slower) in this branch. This may be due to the change of using `Union{Nothing,T}` instead of `Nullable{T}` in some data fields. We still need to diagnose this problem more carefully.

**master (4e609d30911a0658ae743d0e6753b168b9aee805) (Julia 0.6)**:
```
Julia Version 0.6.2
Commit d386e40c17 (2017-12-13 18:08 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
random query
  0.471834 seconds (1.71 M allocations: 56.540 MiB, 2.87% gc time)
  0.227507 seconds (1.56 M allocations: 47.478 MiB, 5.08% gc time)
  0.222676 seconds (1.56 M allocations: 47.478 MiB, 0.82% gc time)
  0.220556 seconds (1.56 M allocations: 47.478 MiB, 0.74% gc time)
  0.211471 seconds (1.56 M allocations: 47.478 MiB, 1.78% gc time)
```

**this pull request (on Julia 0.7-dev)**:
```
Julia Version 0.7.0-DEV.4810
Commit a96d847592* (2018-04-07 21:52 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
Environment:
  JULIA_SHELL = /bin/bash
random query
  1.570659 seconds (16.18 M allocations: 507.332 MiB, 6.57% gc time)
  1.033488 seconds (15.53 M allocations: 467.438 MiB, 1.31% gc time)
  1.016930 seconds (15.53 M allocations: 467.438 MiB, 1.12% gc time)
  1.015666 seconds (15.53 M allocations: 467.438 MiB, 1.18% gc time)
  1.014910 seconds (15.53 M allocations: 467.438 MiB, 1.20% gc time)
```

**benchmark.jl**:
```julia
if VERSION > v"0.7-"
    push!(LOAD_PATH, Base.NamedEnv("v0.7"))
    using InteractiveUtils
    using Random
    using Profile
end
versioninfo()

using IntervalTrees
using CodecZlib

function readgff3(input::IO, chrom::String)
    intervals = Interval{Int}[]
    for line in eachline(input)
        if startswith(line, "#")
            continue
        end
        values = split(line, '\t')
        if values[1] == chrom
            chromstart = parse(Int, values[4])
            chromend = parse(Int, values[5])
            push!(intervals, Interval(chromstart, chromend))
        end
    end
    sort!(intervals)
    return IntervalTree{Int,Interval{Int}}(intervals)
end

function readgff3(filepath::String, chrom::String)
    stream = open(filepath)
    if endswith(filepath, ".gz")
        stream = GzipDecompressorStream(stream)
    end
    try
        return readgff3(stream, chrom)
    finally
        close(stream)
    end
end

# ftp://ftp.ensembl.org/pub/release-92/gff3/danio_rerio/Danio_rerio.GRCz11.92.gff3.gz
tree = readgff3("data/Danio_rerio.GRCz11.92.gff3.gz", "1")

println("random query")
function random_query(tree, queries)
    n = 0
    for q in queries
        for _ in intersect(tree, q)
            n += 1
        end
    end
    return n
end
srand(1234)
intervals = shuffle!(collect(tree))
for _ in 1:5
    @time random_query(tree, intervals)
end
if "-p" in ARGS
    @profile random_query(tree, intervals)
    Profile.print()
end
```

**Comment by sambitdash** [comment]
Can you remove the fast path once and try? This is at a few places. 

         if isbits(T)
-            unsafe_copy!(pointer(xs, i+1),
-                         pointer(xs, i), count - i + 1)
+            unsafe_copyto!(pointer(xs, i+1), pointer(xs, i), count - i + 1)
         else

I felt that was doing a lot number of allocations earlier.

**Comment by bicycle1885** [comment]
No, it has no effect on performance.

```
Julia Version 0.7.0-DEV.4810
Commit a96d847592* (2018-04-07 21:52 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
Environment:
  JULIA_SHELL = /bin/bash
random query
  1.379991 seconds (16.18 M allocations: 507.394 MiB, 6.40% gc time)
  0.967503 seconds (15.53 M allocations: 467.438 MiB, 1.30% gc time)
  1.000985 seconds (15.53 M allocations: 467.438 MiB, 1.10% gc time)
  0.986747 seconds (15.53 M allocations: 467.438 MiB, 1.20% gc time)
  1.064947 seconds (15.53 M allocations: 467.438 MiB, 1.29% gc time)
```

```patch
diff --git a/src/slice.jl b/src/slice.jl
index 66f9703..1408fc4 100644
--- a/src/slice.jl
+++ b/src/slice.jl
@@ -74,13 +74,13 @@ function Base.insert!(s::Slice{T, N}, i::Integer, value) where {T, N}
     if s.n < N && 1 <= i <= s.n + 1
         # TODO: This should work but won't. Fix this in Julia.
         #copy!(s.data, i+1, s.data, i, s.n - i + 1)
-        if isbits(T)
-            unsafe_copyto!(pointer(s.data, i+1), pointer(s.data, i), s.n - i + 1)
-        else
+        #if isbits(T)
+        #    unsafe_copyto!(pointer(s.data, i+1), pointer(s.data, i), s.n - i + 1)
+        #else
             for k in 0:(s.n - i)
                 @inbounds s.data[s.n+1-k] = s.data[s.n-k]
             end
-        end
+        #end
         @inbounds s.data[i] = value
         s.n += 1
         return s
@@ -122,13 +122,13 @@ end
 
 function slice_insert!(xs::Vector{T}, count::Integer, i::Integer, value::T) where T
     if count < length(xs) && 1 <= i <= count + 1
-        if isbits(T)
-            unsafe_copyto!(pointer(xs, i+1), pointer(xs, i), count - i + 1)
-        else
+        #if isbits(T)
+        #    unsafe_copyto!(pointer(xs, i+1), pointer(xs, i), count - i + 1)
+        #else
             for k in 0:(count - i)
                 @inbounds xs[count+1-k] = xs[count-k]
             end
-        end
+        #end
         @inbounds xs[i] = value
         count += 1
         return count
```

**Comment by sambitdash** [comment]
Nullable{T} is no longer in 0.7. There does not seem to be any alternate then, unless we will like to add own version of Nullable{T}.

**Comment by bicycle1885** [comment]
This performance degradation seems to be due to the failure of type inference in Julia. The following patch solves the problem.

```patch
diff --git a/src/IntervalTrees.jl b/src/IntervalTrees.jl
index e2e148a..05c6776 100644
--- a/src/IntervalTrees.jl
+++ b/src/IntervalTrees.jl
@@ -1339,7 +1339,7 @@ function nextintersection!(t::LeafNode{K, V, B}, i::Integer,
 
         t.right === nothing && break
 
-        t = t.right
+        t = t.right::LeafNode{K,V,B}
 
         if minstart(t) > last(query)
             break
```

**9445c93f9754062dfa1f7944dff3e61b9c3d4eb0 + patch**:
```
Julia Version 0.7.0-DEV.4810
Commit a96d847592* (2018-04-07 21:52 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.7.0)
  CPU: Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
Environment:
  JULIA_SHELL = /bin/bash
┌ Warning: `trunc(x::Number, digits)` is deprecated, use `trunc(x; digits=digits)` instead.
│   caller = stale_cachefile(::String, ::String) at loading.jl:1391
└ @ Base loading.jl:1391
random query
  0.747452 seconds (2.00 M allocations: 80.573 MiB, 10.57% gc time)
  0.151449 seconds (1.40 M allocations: 42.690 MiB, 3.20% gc time)
  0.129592 seconds (1.40 M allocations: 42.690 MiB, 1.51% gc time)
  0.128382 seconds (1.40 M allocations: 42.690 MiB, 0.94% gc time)
  0.132978 seconds (1.40 M allocations: 42.690 MiB, 0.89% gc time)
```

**Comment by ScottPJones** [comment]
That's very interesting.  That may in part be why I saw such poor performance with the changes to the `find*` functions to use `nothing` as a sentinel value.

**Comment by bicycle1885** [comment]
The type of `t` is apparent to human beings, but it may not be so to the compiler. This may be a bug of Julia. Adding type annotations is enough now as a workaround.

I think this kind of problems is here and there. We need some a benchmark suite to detect them.

**Comment by sambitdash** [comment]
Or may be just leave the PR for now and have Julia get stable.

**Comment by sambitdash** [comment]
Some.jl in base may have a method to be used. I will look at it tomorrow.

**Comment by sambitdash** [comment]
@ScottPJones, @bicycle1885 `notnothing` does not eliminate the type instability but it ensure no additional allocations are carried out. Comment from the bug states that the `getproperty` has to be put inside a temp variable. I will tend to agree with @bicycle1885 that is an issue the compiler should be able to handle. 

Should we further work on this or wait for the decision of the compiler? 

I will also request you to provide review comments on #36 as well.

**Comment by sambitdash** [comment]
Here is another way to handle the case. Introducing `==(x, ::Nothing)` improves type stability. I guess it's working in the same manner as it's for temp variable or `notnothing`. A multiple dispatch method is forcing the `Union{Nothing, T}` to be evaluated for specific type as the methods are defined for the specific types leading to this behavior. 

```
julia> mutable struct Node{T}
           value::T
           left::Union{Nothing,Node{T}}
           right::Union{Nothing,Node{T}}
           Node(v) = new{typeof(v)}(v, nothing, nothing)
       end

julia> node = Node(1)
Node{Int64}(1, nothing, nothing)

julia> node.right = Node(2)
Node{Int64}(2, nothing, nothing)

julia> import Base: ==

julia> ==(x, ::Nothing) = true
== (generic function with 151 methods)

julia> function rightmost(t::Node)
           while true
               if t.right == nothing
                   break
               end
               t = t.right
           end
           return t
       end
rightmost (generic function with 1 method)

julia> @code_warntype rightmost(node)
Variables:
  t@_2::Node{Int64}

Body:
  begin
      1: 
      #= line 3 =#
      # meta: location sysimg.jl getproperty 18
      Core.SSAValue(2) = (Base.getfield)(t@_2::Node{Int64}, :right)::Union{Nothing, Node{Int64}}
      # meta: pop location
      Core.SSAValue(3) = (Core.SSAValue(2) isa Nothing)::Bool
      unless Core.SSAValue(3) goto 10
      $(Expr(:invoke, MethodInstance for ==(::Nothing, ::Nothing), :(Main.:==), Core.SSAValue(2), :(Main.nothing)))::Bool
      goto 19
      10: 
      Core.SSAValue(4) = (Core.SSAValue(2) isa Node{Int64})::Bool
      unless Core.SSAValue(4) goto 15
      $(Expr(:invoke, MethodInstance for ==(::Node{Int64}, ::Nothing), :(Main.:==), Core.SSAValue(2), :(Main.nothing)))::Bool
      goto 19
      15: 
      goto 17
      17: 
      (Base.error)("fatal error in type inference (type bound)")::Union{}
      19: 
      #= line 4 =#
      goto 23
      #= line 6 =#
      23: 
      #= line 8 =#
      return t@_2::Node{Int64}
  end::Node{Int64}
```

**Comment by sambitdash** [comment]
removed last comment I added as the code was wrong.

**Comment by sambitdash** [comment]
@ScottPJones and @bicycle1885 Do you think we can go ahead with `notnothing`?

**Comment by dcjones** [comment]
This looks good to me.

Should we bump `REQUIRE` to julia 0.7? I would guess that using `nothing` in place of `Nullable` would be such a big performance regression in 0.6, that it might not be worth supporting it at all.

**Comment by ScottPJones** [comment]
Bump `REQUIRE` to v0.7, and merge this?

**Comment by bicycle1885** [comment]
Agreed. The release of Julia 0.7-beta2 is almost there (maybe in a few days?), and so I think it would be better to wait for it and then bump the requirement to 0.7-beta2.

**Comment by phaverty** [comment]
The second beta is out :-)

**Comment by BenJWard** [comment]
Thanks for this @sambitdash and everyone, I'll merge this into the `version-1.0` branch and update the julia REQUIRE version. I'll update the project files and add Documenter.jl docs over on that branch. Thanks for the nice work @sambitdash :+1:

**Comment by sambitdash** [comment]
Thanks a lot for merging the changes.
